gnsi/acltrie/trie.go

```diff
@@ -152,79 +165,104 @@ func (t *Trie) Insert(r *pathzpb.AuthorizationRule) error {
 	return nil
 }
 
-// walk explores the trie in breadth first order and invokes walkFn on every node.
+// Probe returns the action for the given path, user, and mode; if there is no match, deny is returned
+func (t *Trie) Probe(path *gpb.Path, user string, mode pathzpb.Mode) pathzpb.Action {
+	potentialPolicies := []*trieNode{}
+	maxLen := 0
+
+	// Walk the matching rules, keeping only the longest ones that contain either the user or a group that the user belongs too.
+	t.walk(func(node *trieNode, walkPath *gpb.Path) (bool, error) {
+		if res := util.ComparePaths(path, walkPath); res != util.Equal && res != util.Subset {
+			return false, nil
+		}
+		if !node.hasPolicy {
+			return true, nil
+		}
+		if act, _ := node.getAction(user, mode, t.memberships); act != pathzpb.Action_ACTION_UNSPECIFIED {
+			if len(walkPath.Elem) > maxLen {
+				potentialPolicies = nil
+				maxLen = len(walkPath.Elem) - 1
+			}
+			potentialPolicies = append(potentialPolicies, node)
+		}
+
+		return true, nil
+	})
+	if len(potentialPolicies) == 0 {
+		return pathzpb.Action_ACTION_DENY
+	}
+
+	// Pick the policies with the largest number of definite keys.
+	var bestPaths []*trieNode
+	maxSetKeys := -1
+	for _, p := range potentialPolicies {
+		path := &gpb.Path{
+			Elem: make([]*gpb.PathElem, maxLen),
+		}
+		n := p
+		setKey := 0
+		for i := maxLen - 1; i >= 0; i-- {
+			path.Elem[i] = n.elem
+			setKey += setKeys(path.Elem[i].Key)
+			n = p.parent
+		}
+		if setKey >= maxSetKeys {
+			maxSetKeys = setKey
+			bestPaths = append(bestPaths, p)
+		}
+	}
```

Comment:
I wonder if this can be cached in the trie during insertion so there isn't a need for this block?

Reply:
done